include/crypto/event.h

```diff
@@ -0,0 +1,378 @@
+ * dynamically.
+ *
+ * Returns one of these values:
+ * 0    Not set
+ * 1    Set
+ */
+int ossl_event_set0(struct ossl_event_st *event,
+                    uint32_t type, void *ctx,
+                    const void *identifiers,
+                    void *payload, size_t payload_size,
+                    ossl_event_destructor_fn *destructor);
+
+/*
+ * Functions to create and populate event queues
+ * ---------------------------------------------
+ */
+
+/* Allocate / deallocate an event queue dynamically */
+struct ossl_event_queue_st *
+ossl_event_queue_new(const struct ossl_event_queue_method_st *method);
+void ossl_event_queue_free(struct ossl_event_queue_st *queue);
+
+/*
+ * Populate an event queue structure.  Used for event queues that aren't
+ * allocated dynamically.
+ *
+ * Returns one of these values:
+ * 0    Not set
+ * 1    Set
+ */
+int ossl_event_queue_set0(struct ossl_event_queue_st *queue,
+                          const struct ossl_event_queue_method_st *method);
+
+/*
+ * Subscribe / unsubscribe to events on an event queue
+ *
+ * Returns one of these values:
+ * -1   Error
+ *  0   Not found (unsubscibe only)
+ *  1   Subscribed / unsubscribed
+ */
+int ossl_event_queue_subscribe(struct ossl_event_queue_st *queue,
+                               ossl_event_callback_fn *subscriber,
+                               void *subscriber_data);
+
+int ossl_event_queue_unsubscribe(struct ossl_event_queue_st *queue,
+                                 ossl_event_callback_fn *subscriber,
+                                 void *subscriber_data);
+
+/*-
+ * Functions to publish and to wait for events
+ * -------------------------------------------
+ */
+
+/*
+ * Add an event to the queue
+ *
+ * Returns one of these values:
+ *
+ * -1   Error
+ *  0   Event not added (because there's no actual queue)
+ *  1   Event added
+ *
+ * If 0 is returned, an option is to use ossl_event_queue_add_and_wait_for().
+ */
+int ossl_event_queue_add(struct ossl_event_queue_st *queue,
+                         struct ossl_event_st *event);
+
+/*
+ * Remove events from the queue, based on the result of calling the filter
+ *
+ * Returns one of these values:
+ *
+ * -1   Error
+ *  0   There were no events to remove, according to the |filter| call
+ *  1   Events were removed
+ */
+int ossl_event_queue_remove(struct ossl_event_queue_st *queue,
+                            ossl_event_callback_fn *filter,
+                            void *filter_data);
+
+/*
+ * Wait for an event
+ *
+ * This will dispatch all queued events until |subscriber| has been run and
+ * returned something other than |ossl_event_pass|, or until there are no
+ * more events queued in |queue|.
+ *
+ * Returns one of these values:
+ *
+ * -1   Error
+ *  0   |subscriber| was never run, or always returned |ossl_event_pass|
+ *  1   |subscriber| was run, and returned something other than |ossl_event_pass|
+ */
+int ossl_event_queue_wait_for(struct ossl_event_queue_st *queue,
```

Comment:
I'm confused.  Why would the wait_for function need a subscriber argument?  All the subscribers are known by the event queue itself -- and all subscribers will received events as they happen.  What purpose is the subscriber here achieving?  Is it temporarily added to the subscriber list just for this wait?

To me `ossl_event_queue_dispatch` looks like the right API for handling events.  If something needs to specifically wait for it's event, it should flag this in the subscriber data IMO.  I'm rather dubious about the need to wait for _my_ event -- that kind of implies synchronous operation which is not what we want.  We need to support single thread, asynchronous operation, upon this we can build multi-threaded and synchronous modes.

Reply:
I would have no problem throwing this away if that's more desirable.

I had this to support a pull model for events, but if that's deemed never necessary, I'm fine with that too

Reply:
Okay, makes sense.  Why don't we wait and see if this appears like it would be useful before implementing it?

Reply:
Er, are you asking me to remove what I've already implemented?

Reply:
If we're not going to use it, then yes.
That can happen later of course.